unison-core/src/Unison/ABT.hs

```diff
@@ -650,17 +645,31 @@ orderedComponents bs0 = tweak =<< orderedComponents' bs0 where
 
 -- Hash a strongly connected component and sort its definitions into a canonical order.
 hashComponent ::
+  forall a f h v.
   (Functor f, Hashable1 f, Foldable f, Eq v, Show v, Ord v, Ord h, Accumulate h)
   => Map.Map v (Term f v a) -> (h, [(v, Term f v a)])
 hashComponent byName = let
   ts = Map.toList byName
-  embeds = [ (v, void (transform Embed t)) | (v,t) <- ts ]
-  vs = fst <$> ts
-  tms = [ (v, absCycle vs (tm $ Component (snd <$> embeds) (var v))) | v <- vs ]
-  hashed  = [ ((v,t), hash t) | (v,t) <- tms ]
-  sortedHashed = sortOn snd hashed
-  overallHash = Hashable.accumulate (Hashable.Hashed . snd <$> sortedHashed)
-  in (overallHash, [ (v, t) | ((v, _),_) <- sortedHashed, Just t <- [Map.lookup v byName] ])
+  -- First, compute a canonical hash ordering of the component, as well as an environment in which we can hash
+  -- individual names.
+  (hashes, env) = doHashCycle [] ts
+  -- Construct a list of tokens that is shared by all members of the component. They are disambiguated only by their
+  -- name that gets tumbled into the hash.
+  commonTokens :: [Hashable.Token h]
+  commonTokens = Hashable.Tag 1 : map Hashable.Hashed hashes
+  -- Use a helper function that hashes a single term given its name, now that we have an environment in which we can
+  -- look the name up, as well as the common tokens.
+  hashName :: v -> h
+  hashName v = Hashable.accumulate (commonTokens ++ [Hashable.Hashed (hash' env (var v :: Term f v ()))])
+  (hashes', permutedTerms) =
+    ts
+      -- Pair each term with its hash
+      & map (\t -> (hashName (fst t), t))
+      -- Sort again to get the final canonical ordering
+      & sortOn fst
```

Comment:
I'm curious whether we need this sortOn here; `doHashCycle` already returns the component pieces in the canonical right? Do we need the "final" canonical ordering to be different from the canonical ordering returned by `doHashCycle`?

Or maybe is this necessary to keep the same hashes we have on trunk since using the `doHashCycle` ordering would change the hash?

What do you think? 😄

Reply:
I'm pretty sure you have the same thought as I had, can you read the "a little bit too much hashing?" part of the PR summary and let me know if we are talking about the same potential improvement?

Reply:
Ooooh, yeah I didn't realize the "little bit too much hashing" was providing an additional solution, I'll check that out 👍🏼

Reply:
Yes I believe I'm suggesting the same thing you are 👍🏼